### DISCUSSION TITLE
Performance issues in repos with lots of branches

### DISCUSSION THREAD

**Discussion by Tratcher** [gitextensions/gitextensions#11441] [Q&A]
I was working in a repo used for a developer class recently and they hadn't cleaned up everyone's branches for a while, there were hundreds. This caused the git extensions visual branch display to lag horribly and made the program hard to use. I've heard similar feedback about other heavily used repos.

Can anything be done to improve the performance of the app in this scenario? I'm aware of various mitigations like forking with only the primary branch, that's not what I'm after here (though feel free to leave recommendations for others). Hopefully there'd be a way to scale the app to lots of branches without causing a lot of lag.

**Comment by gerhardol** [comment]
Which version did you use? There were many improvements in 4.0 for this scenario.
I have hundreds of branches mysely, not seeing any issue.

You can of course choose to see current branch only by default (and switch to all branches if needed).

Also, what was the exact problem: scrolling? loading?

**Reply by Tratcher** [reply]
Both loading and scrolling were very slow, often pausing for a minute or more.

I counted and found the repo has 13324 branches 😆.

Version:
- Git Extensions 4.2.1.17611
- Build b0c0b284876b63c52c77195ada82a85bfadbb083
- Git 2.43.0.windows.1
- Microsoft Windows NT 10.0.22631.0
- .NET 6.0.25
- DPI 144dpi (150% scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Reply by Tratcher** [reply]
Unfortunately I can't share the repo, it's used for internal training.

**Reply by mstv** [reply]
>  I counted and found the repo has 13324 branches

Does the graph have any benefit in such a repo? You can hide it:

![image](https://github.com/gitextensions/gitextensions/assets/36601201/68f8afae-dcea-46e3-8b64-99d73adc4639)

Otherwise, consider

> You can of course choose to see current branch only by default (and switch to all branches if needed).

![image](https://github.com/gitextensions/gitextensions/assets/36601201/5d1a5174-f7d2-4dcf-94c7-84ffa866f8fc)

What do you miss when using these options?

**Reply by Tratcher** [reply]
Ooo, turning off the graph helps a lot, thanks.

I don't want to just view the current branch, I still need to work with some of the others.

**Reply by mstv** [reply]
Filter with wildcard?

![image](https://github.com/gitextensions/gitextensions/assets/36601201/6db6c5b3-0ac4-4e10-bae1-0d3315989fc9)

**Reply by RussKie** [reply]
> I counted and found the repo has 13324 branches 😆.

🤯 

If you don't need the full history, you could try limiting the number of loaded commits (though this settings is app-global):
![image](https://github.com/gitextensions/gitextensions/assets/4403806/30eb234a-d598-4627-8367-74314dd74ba4)

**Reply by vbjay** [reply]
Should be able to generate a script that generates commit then new branch and new commit...wash rinse repeat...till you have a repo like that.

**Reply by pmiossec** [reply]
>I counted and found the repo has 13324 branches 😆.

I don't know how big is this company but it should be big enough to have this number of branches and no one ever told them that not deleting branches is a bad practice (even for human!)? A branch is to handle something evolving. They should be deleted once merged and they should use tags for ref that no more change...

FWIW, it's not too difficult to generate a big number of branches. I took GE repo and have created 20715 branches (one for each commit by adding carefully adding lines to the `.git/packed-refs` file).
Result is that the loading/refresh is not very quick (around 13s) but once it is loaded, there is no problem and the display and scrolling is snappy. So the problem is not **just** the number of refs but more the resulting complexity of the graph (because in my case the complexity has not been increased).

And in this typo of repositories, hiding the graph is surely the solution.

**Comment by mstv** [comment]
We have added graph tweaks step by step, and have noticed performance issues e.g. with the Linux repo, too (#11292 and disabled some). Which version are you referring to? Can you share an example repo?

**Reply by mstv** [reply]
How does the improved development version https://ci.appveyor.com/api/buildjobs/xbsrvmibr1589cy7/artifacts/artifacts%2FRelease%2Fpublish%2FGitExtensions-Portable-4.3.0.17643-9de19de3e.zip (a build of #11457) work for you, @Tratcher?

**Reply by Tratcher** [reply]
That version performs about the same.

Here's an example where it struggles to scroll:
<img width="234" alt="image" src="https://github.com/gitextensions/gitextensions/assets/1821173/b813efee-411c-4a42-8fd2-e602eb4c2c45">

**Reply by mstv** [reply]
Thank you for testing!
You could add
```
  <item>
    <key>
      <string>Appearance.RevisionGraph.StraightenGraphSegmentsLimit</string>
    </key>
    <value>
      <string>0</string>
    </value>
  </item>
```
to the `GitExtensions.settings` file (in the folder where you unpacked the portable version).

**Reply by Tratcher** [reply]
No effect. The issue primarily happens when scrolling and it has to add or remove columns in the graph display. E.g. every time the branching makes the graph wider or narrower and it has to bump the description over. It also seems to be worse when further back in the history (e.g. scroll down 1000 commits).

**Reply by mstv** [reply]
The loading of the grid has been optimized and the rendering pipeline has been rewritten. Release 5.0 is being prepared.

**Comment by gerhardol** [comment]
I have 650+60 with OK performance in my local GE repo.
At work, branches are cleaned up a lot more aggressive. A reason why -ff-only is the only allowed merge strategy.

But 13K branches will be hard to handle, also if rendering is optimized, if rendering is done in half the time, it would be a big achievement.
So branches will have to purged somehow. I do not have a suggestion on how such a purge can be done automatically. If heads are limited, there will still be a lot of processing still if git-log still emitts all branches.

So you need to reduce the number of branches.

A simple filter is by date or number of commits in advanced filter.

There are more advanced alternatives.
https://git-extensions-documentation.readthedocs.io/en/release-4.0/browse_repository.html#filter-history
GE allows git-log options, adding in the right box, all that start with -- are handled as options. 
See https://git-scm.com/docs/git-log

You may use Current branch together with --glob=refs/remote/origin/foo --refs/remote/origin/bar
The options are persisted, the editor is not great though.

**Comment by vbjay** [comment]
Another small improvement would be to repack refs. https://git-scm.com/docs/git-pack-refs
